File: Python/3_LongestSubstringWoRC.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        """ Brute Force Solution, o(n^2) time sliding window without hashing. Very Slow"""
        if len(s) == 0:
            return 0
        if len(s) == 1:
            return 1

        result = 0 
        l, r  = 0, 1
        sub_string = s[l]

        while r < len(s):
            if s[r] not in sub_string:
                sub_string += s[r]
                r += 1
            else:
                l += 1 
                r = l + 1
                sub_string = s[l]

            result = max(result, (r - l))
        
        return result
    
class Solution2:
    def lengthOfLongestSubstring(self, s: str) -> int:
        """ Sliding Window with Hash, o(n) time o(m) """
        result = 0 
        l = 0
        unique_chars = set()

        for r in range(len(s)):
            while s[r] in unique_chars:
                unique_chars.remove(s[l])
                l += 1
            unique_chars.add(s[r])
            curr = len(unique_chars)
            
            result = max(result, curr)
        
        return result
    
s = Solution2()

print(s.lengthOfLongestSubstring("abcabcbb"))
print(s.lengthOfLongestSubstring("bbbbb"))
print(s.lengthOfLongestSubstring("pwwkew"))
print(s.lengthOfLongestSubstring(""))
print(s.lengthOfLongestSubstring(" "))
print(s.lengthOfLongestSubstring("dvdf"))

File: JavaScript/3_LongestSubstringWoRC.js
/**
 * @param {string} s
 * @return {number}
 */

const lengthOfLongestSubstring = function(s) {
    let result = 0;
    let l = 0;
    const unique_chars = new Set();
    
    for (let r = 0; r < s.length; r++) {
        while (unique_chars.has(s[r])) {
            unique_chars.delete(s[l]);
            l++;
        }
        unique_chars.add(s[r]);
        const currentSubstringLength = unique_chars.size;
        result = Math.max(result, currentSubstringLength);
    }

    return result;
};


const testCases = [
    {
        s: "abcabcbb",
        expected: 3
    },
    {
        s: "bbbbb",
        expected: 1
    },
    {
        s: "pwwkew",
        expected: 3
    },
    {
        s: "",
        expected: 0
    },
    {
        s: " ",
        expected: 1
    },
    {
        s: "dvdf",
        expected: 3
    }
]

testCases.forEach((testCase, i) => {
    const output = lengthOfLongestSubstring(testCase.s);
    if (JSON.stringify(output) === JSON.stringify(testCase.expected)) {
        console.log(`Test Case ${i}: Passed. Expected: ${JSON.stringify(testCase.expected)}, Output: ${JSON.stringify(output)} `)
    } else {
        console.log(`Test Case ${i}: Failed. Expected: ${JSON.stringify(testCase.expected)}, Output: ${JSON.stringify(output)} `)
    }
});